service/dap/server.go

```diff
@@ -2667,10 +2668,170 @@ func (s *Server) onReadMemoryRequest(request *dap.ReadMemoryRequest) {
+// onDisassembleRequest handles 'disassemble' requests.
+// Capability 'supportsDisassembleRequest' is set in 'initialize' response.
 func (s *Server) onDisassembleRequest(request *dap.DisassembleRequest) {
-	s.sendNotYetImplementedErrorResponse(request.Request)
+	// TODO(suzmue): microsoft/vscode#129655 is discussing the difference between
+	// memory reference and instructionPointerReference, which are currently
+	// being used interchangeably by vscode.
+	addr, err := strconv.ParseInt(request.Arguments.MemoryReference, 0, 64)
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToDisassemble, "Unable to disassemble", err.Error())
+		return
+	}
+
+	start := uint64(addr)
+	maxInstructionLength := s.debugger.Target().BinInfo().Arch.MaxInstructionLength()
+	offset := request.Arguments.InstructionOffset * maxInstructionLength
+	// Adjust the offset to include instructions before the requested address.
+	if offset < 0 {
+		start = uint64(addr + int64(offset))
+	}
+	// Adjust the number of instructions to include enough instructions after
+	// the requested address.
+	count := request.Arguments.InstructionCount
+	if offset > 0 {
+		count += offset
+	}
+	end := uint64(addr + int64(count*maxInstructionLength))
+
+	// Make sure the PCs are lined up with instructions.
+	start, end = alignPCs(s.debugger.Target().BinInfo(), start, end)
+
+	// Disassemble the instructions
+	procInstructions, err := s.debugger.Disassemble(-1, start, end)
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToDisassemble, "Unable to disassemble", err.Error())
+		return
+	}
+
+	// Find the section of instructions that were requested.
+	startIdx, endIdx, err := findRange(procInstructions, addr, (request.Arguments.InstructionOffset), (request.Arguments.InstructionCount))
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToDisassemble, "Unable to disassemble", err.Error())
+		return
+	}
+
+	// Turn the given range of instructions into dap instructions.
+	instructions := make([]dap.DisassembledInstruction, endIdx-startIdx)
+	lastFile, lastLine := "", -1
+	for i := 0; i < len(instructions); i++ {
+		instruction := api.ConvertAsmInstruction(procInstructions[startIdx+i], s.debugger.AsmInstructionText(&procInstructions[startIdx+i], proc.GoFlavour))
+		instructions[i] = dap.DisassembledInstruction{
+			Address:          fmt.Sprintf("%#x", instruction.Loc.PC),
+			InstructionBytes: fmt.Sprintf("%x", instruction.Bytes),
+			Instruction:      instruction.Text,
+		}
+		// Only set the location on the first instruction for a given line.
+		if instruction.Loc.File != lastFile || instruction.Loc.Line != lastLine {
+			instructions[i].Location = dap.Source{Path: instruction.Loc.File}
+			instructions[i].Line = instruction.Loc.Line
+			lastFile, lastLine = instruction.Loc.File, instruction.Loc.Line
+		}
+	}
+	response := &dap.DisassembleResponse{
+		Response: *newResponse(request.Request),
+		Body: dap.DisassembleResponseBody{
+			Instructions: instructions,
```

Comment:
The spec says:

> An adapter must return exactly this number of instructions - any
> unavailable instructions should be replaced with an implementation-defined
> 'invalid instruction' value.

I think this can return fewer.

Reply:
What do you mean? The specification specifically says "exactly this number of instructions".

Reply:
My initial implementation returned < request.Arguments.InstructionCount if we reached the end of the region that contains instructions for the program. It has been updated to always return request.Arguments.InstructionCount.